### ISSUE TITLE
Possible codebase restructure

### ISSUE THREAD

**Issue by younik** [GFNOrg/torchgfn#221]
I propose here an alternative codebase organization, with:

`gfn.core`: implementing the interfaces of States, Action, GFlowNet base file, base Env class, Trajectories, Transitions, and ReplayBuffer.
`gfn.discrete`: depends on `gfn.code` interfaces, and implements the DiscreteStates, DiscreteEnv and everything that is specifically to Discrete, like FMGFlowNet.
`gfn.graph` Similarly to before, everything that is related to Graph.

**Motivation:**
Currently, we are mixing different state-type implementations in the same file creates unnecessary dependencies and reduces modularity. Same for environments.
For example, for Graph, I need torch_geometric, but a user may want to use the library only for discrete env and not install torch_geometric.

Dividing in submodules increases modularity and makes it easier to scale the codebase to different use cases

**Comment by younik** [comment]
> Where would classes used for the continuous examples belong in this new structure ?

Good point; box and line directly sublcass Env, and doesn't need special States. We may leave them in `gnf.core` or create a `gfn.continuous`

I am also wondering if we will introduce new Action types in the future that can be mixed with other State types. This will make the submodules confusing.

**Comment by josephdviviano** [comment]
I think it's an interesting idea but we will need to be very careful. The
last restructuring took us a few months and introduced lot of changes.

My hunch is we should hold off on it until torchgfn is more feature
complete, next year. The reason being it will be easier for us to see the
most natural restructuring when we've built everything out.

That might mean the refactoring is a bit more work but it will also be more
likely to be done correctly

Joseph (Mobile)
